### PR TITLE
Add support for local storage feature

### PR DESCRIPTION
**Pull request by phndiaye**
### What does this PR do?

Add support for local storage on the table.

Related to : https://github.com/upfluence/backlog/issues/832

### 🧑‍💻 Developer Heads Up

⚡ Since we are using [Ember Octane](https://blog.emberjs.com/octane-is-here/) now:
* Feel free to migrate existing components to Glimmer Components.
* Write new ones exclusively in it.

Useful Resource : [Ember Octane vs Classic Cheat Sheet](https://ember-learn.github.io/ember-octane-vs-classic-cheat-sheet/)

### Good PR checklist

- [x] Title makes sense
- [x] Is against the correct branch
- [x] Only addresses one issue
- [x] Properly assigned
- [ ] Added/updated tests
- [ ] Added/updated documentation
- [ ] Migrated touched components to Glimmer Components
- [x] Properly labeled

### CODE REVIEW

**Comment by tgallice** [comment]
Having an abstration on the way to store the state of the hypertable could be more usefull, no ? This will let's us to implement a custom table storage if needed.

File: addon/types/table.js [hunk 1]
```diff
@@ -242,5 +244,36 @@ export default EmberObject.extend({
...
+  retrieveLocalState() {
+    if (!this.options.features.localStorage) {
+      throw new Error('[Hypertable] Trying to use localStorage without the feature being enabled.');
+    }
+
+    if (!this.options.name) {
+      throw new Error('[Hypertable] Trying to use localStorage without options.name being set.');
+    }
+
+    console.log(this.options.name);
```
Comment (tgallice) [inline addon/types/table.js:272]: :see_no_evil:

**Review by Hegazy360 — APPROVED** [review]

File: addon/types/store.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    this.table = table;
+    this.key = key;
+  }
+
+  getState() {
+    if (window.localStorage.getItem(this.key)) {
+      return JSON.parse(window.localStorage.getItem(this.key));
+    } else {
+      return null;
+    }
```
Comment (tgallice) [inline addon/types/store.js:12]:
```suggestion
    let state = window.localStorage.getItem(this.key);
   
    return state ? JSON.parse(state) : null;
```

File: addon/types/store.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+class LocalStorageStore {
+  constructor(table, key) {
+    this.table = table;
```
Comment (tgallice) [inline addon/types/store.js:3]: we really want to have the table here ?

Comment (phndiaye) [reply addon/types/store.js:3]: it's to have access to the columns "live-changes"

Comment (tgallice) [reply addon/types/store.js]: Yes, but passing the `columns` to the `update` as argument could be more cleaner, no ?

Comment (phndiaye) [reply addon/types/store.js]: hmm.. you're right :) will update it

**Review by Hegazy360 — APPROVED** [review]

File: addon/types/table.js [hunk 4]
```diff
@@ -55,6 +59,19 @@ export default EmberObject.extend({
...
+    this._super(...arguments);
+
+    if (this.options?.features?.localStorage) {
+      if (!this.options.name) {
+        throw new Error(
+          '[Hypertable] Trying to use localStorage store without a store name. Please set "options.name".'
+        );
+      }
+
+      this.set('store', new LocalStorageStore(this.options.name));
```
Comment (AlexisMontagne) [inline addon/types/table.js:72]: i don't see that being used anywhere? when do you actually interact with the store?

**Comment by AlexisMontagne** [comment]
I do agree with @tgallice we should standardize a way of interacting with the storage of table in hypertable. We have two type of backend handle, and there is two different mechanism.

The CRM with the HTTP backend uses a public hook  to update the table.

I think having this two methods storage interface  is a good start (get, update), and we should make the current table state a private variable, as well as migrate the crm http storage  to a valid storage implementation.

File: addon/types/table.js (hunk 4 above)
Comment (Miexil) [reply addon/types/table.js:72]: I think this will be done in the parent (project, component, etc) scope, where the component is actually used ?

**Review by Miexil — APPROVED** [review]

**Comment by phndiaye** [comment]
@AlexisMontagne the current implementation sets `store` only if `localStorage` is enabled. but in the future, what i had in mind is that the CRM (for example) will be be able to pass an `HTTPStore` that implements the `Store` interface and does the right calls to read and write from/to the API.

**Pull request merged**